core/src/main/java/com/linecorp/armeria/internal/common/BodySubscriber.java

```diff
@@ -0,0 +1,301 @@
+            completeTrailers(trailers);
+            return;
+        }
+
+        assert httpObject instanceof HttpData;
+
+        final EventExecutor executor = downstreamExecutor;
+        if (executor.inEventLoop()) {
+            onNext0((HttpData) httpObject);
+        } else {
+            executor.execute(() -> onNext0((HttpData) httpObject));
+        }
+    }
+
+    private void onNext0(HttpData httpData) {
+        wroteAny = true;
+        if (!usePooledObject) {
+            httpData = PooledObjects.copyAndClose(httpData);
+        }
+        downstream.onNext(httpData);
+    }
+
+    @Override
+    public void onComplete() {
+        maybeCompleteHeaders(null);
+        final EventExecutor executor = downstreamExecutor;
+        final Subscriber<? super HttpData> downstream = this.downstream;
+        if (executor == null || downstream == null) {
+            completing = true;
+            return;
+        }
+
+        if (executor.inEventLoop()) {
+            onComplete0(downstream);
+        } else {
+            executor.execute(() -> onComplete0(downstream));
+        }
+    }
+
+    private static void onComplete0(Subscriber<? super HttpData> downstream) {
+        downstream.onComplete();
+    }
+
+    @Override
+    public void onError(Throwable cause) {
+        maybeCompleteHeaders(cause);
+        final EventExecutor executor = downstreamExecutor;
+        final Subscriber<? super HttpData> downstream = this.downstream;
+        if (executor == null || downstream == null) {
+            this.cause = cause;
+            return;
+        }
+
+        if (executor.inEventLoop()) {
+            onError0(cause, downstream);
+        } else {
+            executor.execute(() -> onError0(cause, downstream));
+        }
+    }
+
+    private void onError0(Throwable cause, Subscriber<? super HttpData> downstream) {
+        downstream.onError(cause);
+        this.downstream = NoopSubscriber.get();
+    }
+
+    /**
+     * Completes the specified trailers.
+     */
+    private void completeTrailers(HttpHeaders trailers) {
+        HeadersFuture<HttpHeaders> trailersFuture = this.trailersFuture;
+        if (trailersFuture != null) {
+            trailersFuture.doComplete(trailers);
+            return;
+        }
+
+        trailersFuture = new HeadersFuture<>();
+        if (trailersFutureUpdater.compareAndSet(this, null, trailersFuture)) {
+            trailersFuture.doComplete(trailers);
+        } else {
+            this.trailersFuture.doComplete(trailers);
+        }
+    }
+
+    protected void maybeCompleteHeaders(@Nullable Throwable cause) {
+        if (trailersFuture == null) {
+            if (trailersFutureUpdater.compareAndSet(this, null, EMPTY_TRAILERS_FUTURE)) {
+                return;
+            }
+        }
+
+        final HeadersFuture<HttpHeaders> trailersFuture = this.trailersFuture;
+        assert trailersFuture != null;
+        trailersFuture.doComplete(HttpHeaders.of());
```

Comment:
This block needs to be called always. Could we make this a private method and provide another method as an extension point?
```java
private void maybeCompleteTraiers() {
    // Complete trailersFuture
    ...
}

// Allow sub-classes to override this.
protected void doOnComplete(@Nullable Throwable cause) {}
```

Reply:
Do you mean fix like this? https://github.com/line/armeria/pull/3953/commits/9884911246f3d02a2da6ce75907cc7267cf7d58d

Reply:
Actually, what I meant was:
```java
class SplitHttpMessageSubscriber {
    ...

    public void onComplete() {
	    doOnCompletion(null);
	    maybeCompleteTrailers();
    }
	  
	public void onError(Throwable cause) {
	   doOnCompletion(cause);
	   maybeCompleteTrailers();
	}
	  
	protected void doOnCompletion(@Nullable Throwable cause) {}
}

class SplitHttpResponseBodySubscriber {
    // No mandatory to call super.doOnCompletion(). 
    @Override
    protected void doOnCompletion(@Nullable Throwable cause) {
        if (!headersFuture.isDone()) {
            if (cause != null && !(cause instanceof CancelledSubscriptionException) &&
                !(cause instanceof AbortedStreamException)) {
                headersFuture.doCompleteExceptionally(cause);
            } else {
                headersFuture.doComplete(HEADERS_WITH_UNKNOWN_STATUS);
            }
        }
    }
}
```

Reply:
Ah, I understood :) And I fixed it in here https://github.com/line/armeria/pull/3953/commits/a51705554e868081e4c9b2cecb4c6a4417abe010
Thank you for the good comments! 👍